iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/parquet/GpuMultiThreadIcebergParquetReader.scala

```diff
@@ -0,0 +1,162 @@
+
+  override def hasNext: Boolean = {
+    ensureDataIterator()
+    if (curDataIterator == null) {
+      false
+    } else {
+      curDataIterator.hasNext
+    }
+  }
+
+  override def next(): ColumnarBatch = {
+    curDataIterator.next()
+  }
+
+  private def ensureDataIterator(): Unit = {
+    if (curDataIterator == null || !curDataIterator.hasNext) {
+      curDataIterator = null
+      if (fileIterator.hasNext) {
+        val file = fileIterator.next()
+        val filePath = file.urlEncodedPath
+        val fileDataIterator = new SingleFileColumnarBatchIterator(filePath,
+          lastBatchHolder, reader, postProcessors)
+        curDataIterator = fileDataIterator
+      }
+    }
+  }
+
+  private def createParquetReader() = {
+    val sparkPartitionedFile = files.map(_.sparkPartitionedFile).toArray
+
+    inited = true
+    new MultiFileCloudParquetPartitionReader(conf.conf,
+      sparkPartitionedFile,
+      this.filterBlock,
+      conf.caseSensitive,
+      conf.parquetDebugDumpPrefix,
+      conf.parquetDebugDumpAlways,
+      conf.maxBatchSizeRows,
+      conf.maxBatchSizeBytes,
+      conf.targetBatchSizeBytes,
+      conf.maxGpuColumnSizeBytes,
+      conf.useChunkedReader,
+      conf.maxChunkedReaderMemoryUsageSizeBytes,
+      CpuCompressionConfig.disabled(),
+      conf.metrics,
+      new StructType(), // iceberg handles partition value by itself
+      conf.threadConf.asInstanceOf[MultiThread].numThreads,
+      conf.threadConf.asInstanceOf[MultiThread].maxNumFilesProcessed,
+      false, // ignoreMissingFiles
+      false, // ignoreCorruptFiles
+      false, // useFieldId
+      // We always set this to true to disable combining small files into a larger one
+      // as iceberg's parquet file may have different schema due to schema evolution.
+      true, // queryUsesInputFile
+      true, // keepReadsInOrder, this is required for iceberg
+      CombineConf(-1, -1)) // Disable combine
+  }
+
+  private def filterBlock(f: PartitionedFile) = {
+    val path = f.filePath.toString()
+    val icebergFile = pathToFile(path)
+
+    val requiredSchema = conf.expectedSchema
+
+    val filteredParquet = super.filterParquetBlocks(icebergFile, requiredSchema)
+
+    val postProcessor = new GpuParquetReaderPostProcessor(
+      filteredParquet.schema,
+      constantsProvider(icebergFile),
+      requiredSchema)
+
+    postProcessors.put(path, postProcessor)
+    filteredParquet
+  }
+}
+
+private class SingleFileColumnarBatchIterator(val targetPath: String,
+    lastBatchHolder: Array[Option[ColumnarBatch]],
+    inner: PartitionReader[ColumnarBatch],
+    postProcessors: ConcurrentMap[String, GpuParquetReaderPostProcessor])
+    extends Iterator[ColumnarBatch]  {
+
+  private def lastBatch: Option[ColumnarBatch] = lastBatchHolder(0)
+
+  override def hasNext: Boolean = if (lastBatch.isEmpty) {
+    if (inner.next()) {
+      lastBatchHolder(0) = Some(inner.get())
+      if (InputFileUtils.getCurInputFilePath() != targetPath) {
```

Comment:
InputFileBlockHolder in Spark is thread-local. Could we here use `getCurInputFilePath` to check if current file is done?

The threads started in `MultiFileCloudParquetPartitionReader` are in a thread pool, and are reading data and set input file path.  The `hasNext` is running in other thread, so could this thread read the expected input file?

```scala
  /**
   * Sets the thread-local input block.
   */
  def set(filePath: String, startOffset: Long, length: Long): Unit = {

```

Reply:
Thanks for your explaination.